clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -154,6 +158,8 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     private AtomicBoolean fencedInstance = new AtomicBoolean(false);
     private final Optional<String> groupInstanceId;
 
+    private ExecutorService callbackExecutor = Executors.newSingleThreadExecutor();
```

Comment:
This is a bit weird, no? In the current model, I suppose that the callbacks are executed by the application thread when poll() is called. Should we do the same here?

Reply:
Sorry - this is wrong.  The way I wanted to do it in this PR is to register the callbacks, when the future is completed, to the application thread.  And we invoke them when we call poll.  I believe the callback invocation is currently invoked on consumer.poll(), because this is where network clients are polled.